src/test/java/com/google/googleidentity/oauth2/validator/TokenEndpointRequestValidatorTest.java

```diff
@@ -0,0 +1,401 @@
+/*
+    Copyright 2020 Google LLC
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+package com.google.googleidentity.oauth2.validator;
+
+import com.google.common.base.Charsets;
+import com.google.common.hash.Hashing;
+import com.google.googleidentity.oauth2.client.ClientDetails;
+import com.google.googleidentity.oauth2.client.ClientDetailsService;
+import com.google.googleidentity.oauth2.client.ClientSession;
+import com.google.googleidentity.oauth2.client.InMemoryClientDetailsService;
+import com.google.googleidentity.oauth2.exception.*;
+import com.google.googleidentity.oauth2.util.OAuth2Constants;
+import com.google.googleidentity.oauth2.util.OAuth2ParameterNames;
+import com.google.googleidentity.security.UserSession;
+import com.google.googleidentity.testtools.FakeHttpSession;
+import com.google.googleidentity.user.UserDetails;
+import org.junit.Before;
+import org.junit.Test;
+
+import javax.servlet.http.HttpServletRequest;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class TokenEndpointRequestValidatorTest {
+    private ClientSession clientSession;
+
+    private ClientDetailsService clientDetailsService;
+
+
+    private static final String CLIENTID = "111";
```

Comment:
nit: it may help to make all these strings different, so that you can tell if you accidentally mix them up. Otherwise, you won't be able to tell if they were mixed up in your code since the assertions would still pass.